modules/create_timepoint/php/ajaxtimepoint.class.inc

```diff
@@ -0,0 +1,280 @@
+            $site = \Site::singleton(intval($values['psc']));
+        }
+
+        //Project
+        $projects = \Utility::getProjectList();
+        if (count($projects) == 1) {
+            //if there is only one project, autoselect first project from array of 1
+            //TODO: change this to array_key_first() when support is only PHP 7.3+
+            $project = \Project::singleton(array_pop($projects));
+        } else if (count($projects) > 1) {
+            $project_id = intval($values['psc']);
+            $project    = \Project::singleton($projects[$project_id]);
+        }
+
+        assert(isset($site) && $site !== null);
+        $errors   = $values['errors'];
+        $response = array();
+        if (!$errors && $site) {
+            \TimePoint::createNew(
+                \Candidate::singleton($values['identifier']),
+                intval($values['subproject']),
+                $values['visit'],
+                $site,
+                $project
+            );
+            $response['status'] = 'success';
+        } else {
+            $response['status'] = 'error';
+            $response['errors'] = $errors;
+        }
+        return new \LORIS\Http\Response\JsonResponse(
+            $response
+        );
+    }
+    /**
+     * Validate the post data,
+     * from the create timepoint form.
+     *
+     * @param array $values the form values.
+     *
+     * @return array
+     */
+    private function _validate(array $values) : array
+    {
+        $user = \User::singleton();
+        $errors = array();
+        if (!isset($values['identifier'])) {
+            $values['error'] = 'BadRequest';
+            return $values;
+        }
+        $values['identifier'] = isset($values['identifier']) ?
+            intval($values['identifier']) : null;
+        $values['psc'] = empty($values['psc']) ?
+            null : $values['psc'];
+        // validate site entered
+        $site = (int) $values['psc'];
+        $user_list_of_sites = $user->getData('CenterIDs');
+        $num_sites          = count($user_list_of_sites);
+        if ($num_sites > 1 && (empty($site) || !$user->hasCenter($site))) {
+            $errors['psc'] = 'Site must be selected from the available dropdown.';
+        }
+        $candid       = intval($values['identifier']);
+        $subprojectID = intval($values['subproject']);
+        $visitLabel   = $values['visit'] ?? '';
+        try {
+            \TimePoint::isValidVisitLabel($candid, $subprojectID, $visitLabel);
+        } catch (\LorisException $exception) {
+            $errors['visitLabel'] = $exception->getMessage();
+        }
+        $values['errors'] = $errors;
+        return $values;
+    }
+    /**
+     * Permission verification.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     * @param array                  $values  The form values.
+     *
+     * @return array
+     */
+    private function _verifyPermissions(
```

Comment:
Why not use `_hasAccess`?

Reply:
Hey @johnsaigle I'm making a change to include the _hasAccess but I'm still needing the _verifyPermissions function. The new endpoints process for handling the requests was introduced by @xlecours and has the _hasAccess called before the 'get' or 'post' request data can be processed. The permissions needs to check the "candidate" for RegistrationCenterID & CenterID;. The candidate is made with the needed identifier sent from the client.